src/main/java/mesosphere/marathon/core/leadership/CandidateImpl.java

```diff
@@ -121,17 +126,38 @@ public CandidateImpl(
     public Supplier<Boolean> offerLeadership(final Leader leader)
             throws Group.JoinException, Group.WatchException, InterruptedException {
 
-        final Group.Membership membership = group.join(dataSupplier, new Command() {
-            @Override public void execute() {
-                leader.onDefeated();
-            }
-        });
+        // start a group watch, but only once in the life time of the the CandidateImpl.
+        if (this.cancelWatch == null) {
+            this.cancelWatch = this.group.watch(new Group.GroupChangeListener() {
+                @Override
+                public void onGroupChange(Iterable<String> memberIds) {
+                    Group.GroupChangeListener listener = groupChangeListener.get();
+                    if (listener != null) {
+                        listener.onGroupChange(memberIds);
+                    }
+                }
+            });
+        }
 
+        // listen for group member changes
         final AtomicBoolean elected = new AtomicBoolean(false);
         final AtomicBoolean abdicated = new AtomicBoolean(false);
-        group.watch(new Group.GroupChangeListener() {
-            @Override public void onGroupChange(Iterable<String> memberIds) {
+        final AtomicReference<Group.Membership> membershipRef = new AtomicReference<Group.Membership>(null);
+        final AtomicReference<Iterable<String>> pendingChange = new AtomicReference<Iterable<String>>(null);
+        this.groupChangeListener.set(new Group.GroupChangeListener() {
+            @Override
+            public void onGroupChange(Iterable<String> memberIds) {
                 boolean noCandidates = Iterables.isEmpty(memberIds);
+                Group.Membership membership;
+                synchronized (this) {
```

Comment:
It*s not about the ZK objects. It's about synchronizing membershipRef.get and pendingChange.set. I want to avoid that the call to onGroupChange outside of the listener calls the handler with an old value of memberIds.